benches: Add benchmark to test duplicate-inputs worst case scenario

jrakibi opened the pull request:
Related to https://github.com/rust-bitcoin/rust-bitcoin/pull/5402 

I tested the worst-case scenario by putting the duplicate at the very end.

The 3-checks algorithm achieves the following performance for 1000 inputs:

- BTreeSet (existing implementation): `[47.464 µs 47.867 µs 48.289 µs]`

- Sorted list: `[14.585 µs 14.640 µs 14.705 µs]` (about 3.3× faster than BTreeSet)

- Pairwise: `[697.03 ns 707.24 ns 722.44 ns]`

This is about 67× faster than BTreeSet. 

Weirdly enough, pairwise performed better than BTreeSet even for 10,000 inputs (see graph below). I was expecting it to be slower as the size increases. 

*I haven’t included the bucket-sort idea mentioned here https://github.com/rust-bitcoin/rust-bitcoin/pull/5402#issuecomment-3647059295 because of the DoS risk, but if it’s worth adding to this benchmark, I can do it.*

I also ran a benchmark with 10_000 inputs that I didn’t include in this code, and the result is that pairwise is always performing better. The graph below shows the result for 10,000 inputs.  

<img width="1024" height="532" alt="image" src="https://github.com/user-attachments/assets/a5610983-b59d-4e40-addb-45d5f3cfa3c0" />

*I’m not very familiar with Criterion benchmarking, so if anything else is needed, let me know*

Addresses https://github.com/rust-bitcoin/rust-bitcoin/issues/5410

apoelstra commented:
The `pairwise` benchmark is so fast because you've constructed your "worst case" to be very favorable. First, your txids (mostly) all differ in the first byte. Secondly, you've constructed your duplicates so that only `n` checks are needed to find a duplicate, because the 0th input is a duplicate.

If you made the last *two* inputs the same, but all others unique, and have all the unique ones have equal TXIDs except the last byte, this would be a better worst-case comparison.

apoelstra commented:
For your curiosity with 64k inputs I see
```
duplicate_inputs/pairwise/65000
                        time:   [1.5105 s 1.5118 s 1.5132 s]
                        thrpt:  [42.954 Kelem/s 42.995 Kelem/s 43.031 Kelem/s]
duplicate_inputs/btreeset/65000
                        time:   [9.0205 ms 9.0226 ms 9.0249 ms]
                        thrpt:  [7.2023 Melem/s 7.2041 Melem/s 7.2058 Melem/s]
duplicate_inputs/sorted/65000
                        time:   [244.12 µs 244.45 µs 244.84 µs]
                        thrpt:  [265.48 Melem/s 265.90 Melem/s 266.26 Melem/s
```

So the pairwise impl now takes over a *second*, the btreeset is somehow into the millisecond range, and the sorting time is still on the order of a single sha2 compression.

Using `sort_unstable` made no measurable difference for me, and using `sort_unstable_by_key` to sort u16 indices rather than the outpoints themselves resulted in a 2-3x slowdown for large input sets and may not be worthwhile given that we're saving a couple hundred kb of memory on a multi-megabyte transaction.

jrakibi commented:
On my system, it looks like beyond ~5000 inputs Hashset starts to outperforms both sort and sort_unstable

<img width="1710" height="900" alt="image" src="https://github.com/user-attachments/assets/92df6302-0f95-4beb-911d-aaf500366c0d" />

apoelstra commented:
> What do you think about using `HashSet`? It looks like it beats sorting for large n.

It's not available with `alloc`, only `std`. Otherwise we'd be using it everywhere.

jrakibi commented:
That makes sense, I’ll take a look tomo

jrakibi commented:
I added benchmarks with different block sizes: 2500 txs from a real mainnet block, and 1k, 10k, 64k test transactions.

For the 64k tx block, here are the results:

Before adding duplicate input check:

```
large_block/decode/64000tx
                        time:   [18.506 ms 18.584 ms 18.669 ms]
                        
large_block/decode_then_validate/64000tx
                        time:   [98.192 ms 98.689 ms 99.373 ms]
```

After adding the check:

```
large_block/decode/64000tx
                        time:   [19.508 ms 19.752 ms 20.179 ms]

large_block/decode_then_validate/64000tx
                        time:   [99.231 ms 99.439 ms 99.661 ms]
```

> The unchecked->checked conversion for a 64k-transaction block, I would expect to be on the order of 100ms

That's exactly what I'm getting. before the check: 98.7ms, after: 99.4ms. The increase is only ~0.7ms.

> But the decoding itself is probably well under 4ms

For decoding, It takes 2.49ms after adding the check (2.24ms before) for the mainnet block

For a 64k tx block, decode takes ~18.6ms before and ~19.8ms after, so the overhead is ~1.2ms (+6%). Either way I guess it's negligible compared to the total decode + validate time 


Note: To get these before/after results, I manually added and removed the duplicate input check from the transaction decoder. I couldn't think of a way to benchmark both in one run since the check is embedded in the decoder logic

jrakibi commented:
Here are the rest of the benchmark results if needed:

BEFORE adding the check: 
```

decode_and_validate/decode/2500tx
                        time:   [2.2364 ms 2.2460 ms 2.2576 ms]

decode_and_validate/decode_then_validate/2500tx
                        time:   [14.389 ms 14.532 ms 14.762 ms]

large_block/decode/1000tx
                        time:   [249.49 µs 250.97 µs 252.84 µs]

large_block/decode_then_validate/1000tx
                        time:   [1.4981 ms 1.5026 ms 1.5076 ms]

large_block/decode/10000tx
                        time:   [2.6174 ms 2.6263 ms 2.6365 ms]

large_block/decode_then_validate/10000tx
                        time:   [15.128 ms 15.234 ms 15.399 ms]

large_block/decode/64000tx
                        time:   [18.506 ms 18.584 ms 18.669 ms]

large_block/decode_then_validate/64000tx
                        time:   [98.192 ms 98.689 ms 99.373 ms]

```

AFTER

```
decode_and_validate/decode/2500tx
                        time:   [2.4778 ms 2.4853 ms 2.4940 ms]

decode_and_validate/decode_then_validate/2500tx
                        time:   [14.657 ms 14.783 ms 15.000 ms]

large_block/decode/1000tx
                        time:   [263.54 µs 264.44 µs 265.40 µs]

large_block/decode_then_validate/1000tx
                        time:   [1.5089 ms 1.5127 ms 1.5168 ms]

large_block/decode/10000tx
                        time:   [2.7568 ms 2.7668 ms 2.7800 ms]

large_block/decode_then_validate/10000tx
                        time:   [15.286 ms 15.461 ms 15.780 ms]

large_block/decode/64000tx
                        time:   [19.508 ms 19.752 ms 20.179 ms]

large_block/decode_then_validate/64000tx
                        time:   [99.231 ms 99.439 ms 99.661 ms]
```

apoelstra commented:
Based on these benchmarks I say "let's go for it". I do feel that block decoding (without hashing transactions) should be faster but that's a separate issue.

jrakibi commented:
I’m leaning toward merging it. I plan to do a proper investigation of the encoding/decoding impl (old vs new) later this month, and these benchmarks will be very helpful for that work

apoelstra reviewed (approved):
ACK bd8a422960e57d2396b5677c22de3e200d173ce0; successfully ran local tests

tcharding reviewed (approved):
ACK bd8a422960e57d2396b5677c22de3e200d173ce0

tcharding commented:
PR description mentions #5402, did you mean that?

jrakibi commented:
> PR description mentions https://github.com/rust-bitcoin/rust-bitcoin/pull/5402, did you mean that?

Sorry for the confusion, I meant "related to" not "on top of" 5402. this can be merged independently.
Fixed description

apoelstra reviewed (approved):
ACK bd8a422960e57d2396b5677c22de3e200d173ce0; successfully ran local tests
